vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1259 @@
+            loc_ loc = tile_to_loc_[tile_id];
+            auto key1 = redirect_[std::tuple<int, loc_>(node1, loc)];
+            auto key2 = redirect_[std::tuple<int, loc_>(node2, loc)];
+
+            VTR_ASSERT(loc_type_idx_to_rr_idx_.find(key1) != loc_type_idx_to_rr_idx_.end());
+            VTR_ASSERT(loc_type_idx_to_rr_idx_.find(key2) != loc_type_idx_to_rr_idx_.end());
+
+            int src, sink;
+            src = loc_type_idx_to_rr_idx_[key1];
+            sink = loc_type_idx_to_rr_idx_[key2];
+            device_ctx_.rr_graph_builder.emplace_back_edge(RRNodeId(src), RRNodeId(sink), sw_id);
+
+            vtr::interned_string name_(empty_);
+            vtr::interned_string value_(empty_);
+
+            char metadata_[100];
+            sprintf(metadata_, "%d,%d,%d,%d", tile_id, wire0, wire1, forward ? 1 : 0);
+
+            name_ = device_ctx_.arch->strings.intern_string(vtr::string_view("FPGAInterchange"));
+            value_ = device_ctx_.arch->strings.intern_string(vtr::string_view(metadata_));
+            vpr::add_rr_edge_metadata(src, sink, sw_id, name_, value_);
+        }
+    }
+
+    void pack_rr_edges() {
+        pack_tiles_edges();
+        pack_chans_edges();
+        pack_pips();
+        device_ctx_.rr_graph_builder.mark_edges_as_rr_switch_ids();
+        device_ctx_.rr_graph_builder.partition_edges();
+    }
+
+    /*
+     * Fill device rr_graph informations.
+     */
+
+    void pack_to_rr_graph() {
+        pack_chan_width();
+        // switches are already packed
+        // segments are already packed before rr_generation
+        // physical_tile_types are already packed before rr_generation
+        // grid is already packed before rr_generation
+        pack_rr_nodes();
+        pack_rr_edges();
+    }
+
+    void finish() {
+        for (t_rr_type rr_type : RR_TYPES) {
+            if (rr_type == CHANX) {
+                device_ctx_.rr_graph_builder.node_lookup().resize_nodes(grid_.height(), grid_.width(), rr_type, NUM_SIDES);
+            } else {
+                device_ctx_.rr_graph_builder.node_lookup().resize_nodes(grid_.width(), grid_.height(), rr_type, NUM_SIDES);
+            }
+        }
+        for (size_t node = 0; node < device_ctx_.rr_nodes.size(); node++) {
+            auto rr_node = device_ctx_.rr_nodes[node];
+            device_ctx_.rr_graph_builder.add_node_to_all_locs(rr_node.id());
+        }
+
+        device_ctx_.rr_graph_builder.init_fan_in();
+
+        alloc_and_load_rr_indexed_data(
+            segment_inf_,
+            0, //we connect ipins to tracks with shorts
+            base_cost_type_);
+
+        VTR_ASSERT(device_ctx_.rr_indexed_data.size() == seg_index_.size());
+        for (size_t i = 0; i < seg_index_.size(); ++i) {
+            device_ctx_.rr_indexed_data[RRIndexedDataId(i)].seg_index = seg_index_[RRIndexedDataId(i)];
+        }
+    }
+
+    /*
+     * Create site sink/source ipin/opin mappins.
+     */
+
+    void create_sink_source_nodes() {
```

Comment:
I believe that this function might be moved on top with the other functions that prepare the data for the rr_graph_builder.

Reply:
fixed in new commit